[GCI] Experimental GIF Manipulation

HarshKhandeparkar opened the pull request:
Fixes #75 (<=== Replace `0000` with the Issue Number)

![Screenshot from 2019-12-24 00-35-20](https://user-images.githubusercontent.com/34770591/71375932-d042d200-25e5-11ea-8467-e7b8d1bef83f.png)

![gif-manip](https://user-images.githubusercontent.com/34770591/71375854-98d42580-25e5-11ea-8dfe-81d9a7c64fa4.gif)

#### Important
1. Modules that do not use PixelManipulation.js are not working with GIFs but work with still imgs.
2. It seems `save-pixels` throws an error while saving GIFs so I have used gifshot. But, the frame-rate/frame duration of the GIF has been fixed to 1/10 s.
3. The modules listed below as 'not working' don't work for GIFs but work for still images. I will fix those in a follow-up PR.

Modules which work:
1. Average
2. Blur
3. Brightness
4. Channel
5. Color Temperature
6. Color Map
7. Contrast
8. Convolution
9. Dither
10. Dynamic
11. Exposure
12. Gamma Correction
13. Grid Overlay
14. Import Image
15. Paint Bucket
16. Saturation
17. Thresholding
18. Tint
19. Draw Rectangle
20. Flip Image
21. Invert
22. Gradient
23. Replace Color
24. Edge Detect
25. Add QR
26. Crop
27. Rotate
28. Resize
29. Canvas Resize
30. Blob Analysis
31. NDVI
32. NDVI Colormap
33. Noise Reduction


Modules which don't work:
1. ColorBar ( Will get fixed upon fixing overlay as this is a meta module which uses overlay)
2. FisheyeGL (It doesn't use pixel manipulation I think)
3. Minify Image (It was broken earlier)
4. Overlay (Need to add new code to retrieve pixels of earlier steps)
5. Text Overlay (Almost Done)
6. Blend (Need to add new code to retrieve pixels of earlier steps)
7. Histogram
8. WebGL Distort (This one's a bit more difficult than others, but it works for still images)

Modules which I am not sure about:
2. Decode QR? (Should it even be used on a GIF?)

Whew. That was a big list. 
Can somebody confirm the above list or check if the modules that I'm not sure about, are working or not? Thank you!

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `npm test`
* [x] code is in a uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] ask `@publiclab/is-reviewers` for help, in a comment below
* [x] Insert-step functionality is working correctly as expected.
> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software
Please make sure to get at least two reviews before asking for merging the PR as that would make the PR more reliable on our part
Thanks!

jywarren commented on src/modules/EdgeDetect/Module.js:
```diff
@@ -19,42 +20,31 @@ module.exports = function edgeDetect(options, UI) {
+
+    // Extra Manipulation function used as an enveloper for applying gaussian blur and Convolution
+    function extraManipulation(pixels) {
+      const blurPixels = Blur(pixels, options.blur);
```
This is great, but was it just optimization or are we not able to do internal sequences anymore?

keshav234156 replied:
@jywarren I guess it just and optimization and it amazing as well

jywarren replied:
Let's leave this comment open so people know!

jywarren commented:
I see this error:

https://travis-ci.org/publiclab/image-sequencer/jobs/631323562#L681

And, some per module tests, like colorbar and qr code, failing. But were these already failing? We really ought to resolve them (and be sure Travis can detect them failing!) because it makes this much harder to review. 

@keshav234156 @vaibhavmatta what do you think of this PR? 

Thanks all!!!

HarshKhandeparkar commented:
Thanks for the review!
1. What are internal sequences?
2. The tests are failing not because modules are wrong but because tests
are wrong. Some modules were producing wrong outputs earlier which I now
fixed. I didn't fix the tests in this PR because it would be out of the
scope of GCI or this specific issue. There is already another PR open and
it can be merged as soon as this one is merged.
The other PR will fix all the tests. And it seems that tests like Tint and
Rotate were failing a long time ago.
Travis doesn't fail even though these tests fail. Many modules were updated
in between but the corresponding tests were not.

On Wed, 1 Jan, 2020, 9:28 PM Jeffrey Warren, <notifications@github.com>
wrote:

> I see this error:
>
> https://travis-ci.org/publiclab/image-sequencer/jobs/631323562#L681
>
> And, some per module tests, like colorbar and qr code, failing. But were
> these already failing? We really ought to resolve them (and be sure Travis
> can detect them failing!) because it makes this much harder to review.
>
> @keshav234156 <https://github.com/keshav234156> @vaibhavmatta
> <https://github.com/vaibhavmatta> what do you think of this PR?
>
> Thanks all!!!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/image-sequencer/pull/1404?email_source=notifications&email_token=AIJI5H7RCMDD4P7KRTJTISDQ3S4QXA5CNFSM4KBWFPVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH5HVNI#issuecomment-570063541>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIJI5H6AINUYYIYKHIW6AF3Q3S4QXANCNFSM4KBWFPVA>
> .
>

HarshKhandeparkar commented:
I will also add GIF tests in the other PR.

On Wed, 1 Jan, 2020, 9:41 PM M K, <mjkgoa@gmail.com> wrote:

> Thanks for the review!
> 1. What are internal sequences?
> 2. The tests are failing not because modules are wrong but because tests
> are wrong. Some modules were producing wrong outputs earlier which I now
> fixed. I didn't fix the tests in this PR because it would be out of the
> scope of GCI or this specific issue. There is already another PR open and
> it can be merged as soon as this one is merged.
> The other PR will fix all the tests. And it seems that tests like Tint and
> Rotate were failing a long time ago.
> Travis doesn't fail even though these tests fail. Many modules were
> updated in between but the corresponding tests were not.
>
> On Wed, 1 Jan, 2020, 9:28 PM Jeffrey Warren, <notifications@github.com>
> wrote:
>
>> I see this error:
>>
>> https://travis-ci.org/publiclab/image-sequencer/jobs/631323562#L681
>>
>> And, some per module tests, like colorbar and qr code, failing. But were
>> these already failing? We really ought to resolve them (and be sure Travis
>> can detect them failing!) because it makes this much harder to review.
>>
>> @keshav234156 <https://github.com/keshav234156> @vaibhavmatta
>> <https://github.com/vaibhavmatta> what do you think of this PR?
>>
>> Thanks all!!!
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/publiclab/image-sequencer/pull/1404?email_source=notifications&email_token=AIJI5H7RCMDD4P7KRTJTISDQ3S4QXA5CNFSM4KBWFPVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH5HVNI#issuecomment-570063541>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AIJI5H6AINUYYIYKHIW6AF3Q3S4QXANCNFSM4KBWFPVA>
>> .
>>
>

SidharthBansal commented:
Happy new year Jeff and Harsh 🎉 
@jywarren do you think that score of this task should be 15 points instead of 5 points?
I mean should we switch it to `hall of fame` from `hard`?

SidharthBansal commented:
28 files is EPIC

keshav234156 commented:
@jywarren Additionally crop and add-qr failed, crop has also been fixed in this PR and for add-qr, there has been some optimization and it's perfect. And changes also looks good to me except  this error:
https://travis-ci.org/publiclab/image-sequencer/jobs/631323562#L681

HarshKhandeparkar commented:
I'm pretty sure I haven't touched anything related to canvas. Maybe one of
the libraries uses canvas? Wait... Does GifShot use canvas? I'm not sure.
This could screw up everything.

On Wed, 1 Jan, 2020, 10:16 PM keshav234156, <notifications@github.com>
wrote:

> @jywarren <https://github.com/jywarren> Additionally crop and add-qr
> failed, crop has also been fixed in this PR and for add-qr, there has been
> some optimization and it's perfect. And changes also looks good to me
> except this error:
> https://travis-ci.org/publiclab/image-sequencer/jobs/631323562#L681
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/image-sequencer/pull/1404?email_source=notifications&email_token=AIJI5H3NIOF42ABRFLSQICDQ3TCE7A5CNFSM4KBWFPVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH5IOFQ#issuecomment-570066710>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIJI5HZO3EIBT2ZHCZRFEK3Q3TCE7ANCNFSM4KBWFPVA>
> .
>

jywarren commented:
I do think this should be hall of fame 🎉 

Hi all! Don't work too hard today!

Also noting possible relation to documented test failures in https://github.com/publiclab/image-sequencer/issues/1376 ?

Let's doc if GifShot uses canvas. No rush here, no worries, let's just take it one step at a time!!

keshav234156 commented:
@jywarren Lets first fix #1376 .Although in #1389 some of test are fixed but some are still failing .So i will now first fix #1376 by today

jywarren commented:
Sounds super. Thank you!!!

HarshKhandeparkar commented:
@jywarren should I add tests in a separate PR? I think that is better because this one already exceeds 25 files. Also, I want to get the GCI task approved. This one took very long.

jywarren commented:
I'd like to see even a single test with a sample GIF, it could be very
small but just to confirm that GIFs are working at all! I hope that is OK.

On Thu, Jan 2, 2020 at 1:26 PM Harsh Khandeparkar <notifications@github.com>
wrote:

> @jywarren <https://github.com/jywarren> should I add tests in a separate
> PR? I think that is better because this one already exceeds 25 files. Also,
> I want to get the GCI task approved. This one took very long.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/image-sequencer/pull/1404?email_source=notifications&email_token=AAAF6JZD36DLC5GT3WHJHQ3Q3YWVZA5CNFSM4KBWFPVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH7AR6Y#issuecomment-570296571>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAF6J3RJPCWBUBGMXGMGMLQ3YWVZANCNFSM4KBWFPVA>
> .
>

HarshKhandeparkar commented:
Okay. I'll do that tmrw.

On Thu, 2 Jan, 2020, 11:57 PM Jeffrey Warren, <notifications@github.com>
wrote:

> I'd like to see even a single test with a sample GIF, it could be very
> small but just to confirm that GIFs are working at all! I hope that is OK.
>
> On Thu, Jan 2, 2020 at 1:26 PM Harsh Khandeparkar <
> notifications@github.com>
> wrote:
>
> > @jywarren <https://github.com/jywarren> should I add tests in a separate
> > PR? I think that is better because this one already exceeds 25 files.
> Also,
> > I want to get the GCI task approved. This one took very long.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/publiclab/image-sequencer/pull/1404?email_source=notifications&email_token=AAAF6JZD36DLC5GT3WHJHQ3Q3YWVZA5CNFSM4KBWFPVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH7AR6Y#issuecomment-570296571
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAAF6J3RJPCWBUBGMXGMGMLQ3YWVZANCNFSM4KBWFPVA
> >
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/image-sequencer/pull/1404?email_source=notifications&email_token=AIJI5HYXQB3TOZZIYP54DTLQ3YW2XA5CNFSM4KBWFPVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH7AVDQ#issuecomment-570296974>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIJI5H4DDJPAPLOQUZQVKXDQ3YW2XANCNFSM4KBWFPVA>
> .
>

HarshKhandeparkar commented:
@jywarren GIF manipulation works in nodejs now. The library [gifshot](https://github.com/yahoo/gifshot ) requires the `canvas` API and doesn't work in node by default. I had to use puppeteer to make it work.
It's fixed now. Could you please review again? I'll add the test in a few mins.

HarshKhandeparkar commented:
@jywarren I've added the test. We currently don't have a way of comparing GIFs so the test compares the data URIs. This doesn't account for human perception and other technical stuff but is the best solution at the time. The output GIF is saved to `/test_outputs/` and the test doesn't fail.

jywarren commented:
That's fine about the data-url comparison, that's how we do it in other comparison based tests too! (although there is also some module called like "image-diff" or something, kind of interesting)

1. What are internal sequences?

Forgot to answer. This is where a new instance of IS is started within a module, run, and its output it passed back as the output of the module. Not extremely useful but occasionally it makes sense. It does allow for multi-threading although that's not why it's used here.

jywarren commented:
Hmm, the jest tests were failing. Any idea why?

```
FAIL  test/ui-2/test/radio-group.test.js
  Radio Group
    ✕ Resize Module is added (902ms)
    ✕ Brightness Module is added (35ms)
    ✕ Contrast Module is added (23ms)
    ✕ Saturation Module is added (14ms)
    ✕ Rotate Module is added (17ms)
    ✕ Crop Module is added (17ms)
```

HarshKhandeparkar commented:
No idea. I didn't touch the UI.

On Sat, 4 Jan, 2020, 12:51 AM Jeffrey Warren, <notifications@github.com>
wrote:

> Hmm, the jest tests were failing. Any idea why?
>
> FAIL  test/ui-2/test/radio-group.test.js
>
>   Radio Group
>
>     ✕ Resize Module is added (902ms)
>
>     ✕ Brightness Module is added (35ms)
>
>     ✕ Contrast Module is added (23ms)
>
>     ✕ Saturation Module is added (14ms)
>
>     ✕ Rotate Module is added (17ms)
>
>     ✕ Crop Module is added (17ms)
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/image-sequencer/pull/1404?email_source=notifications&email_token=AIJI5H5EGDMONP6VLHKWWZDQ36F3JA5CNFSM4KBWFPVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIB4DWY#issuecomment-570671579>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIJI5H25FZM5YT2XLE7KXNDQ36F3JANCNFSM4KBWFPVA>
> .
>

HarshKhandeparkar commented:
I'm not sure what's happening to the tests but I am pretty sure it has got something to do with the new function I added that uses puppeteer to save GIFs in nodejs.

HarshKhandeparkar commented:
I'll try to find out what's happening tomorrow.

HarshKhandeparkar commented:
Even the GIF tests are passing 
```
  GIF input works correctly
    ✔ should be equal
  total:     1
  passing:   1
  duration:  2.4s
```
but the UI tests are failing. I think I will need @keshav234156 's help in this.
```
 FAIL  test/ui-2/test/radio-group.test.js (5.127s)
  Radio Group
    ✕ Resize Module is added (965ms)
    ✕ Brightness Module is added (2248ms)
    ✕ Contrast Module is added (30ms)
    ✕ Saturation Module is added (28ms)
    ✕ Rotate Module is added (29ms)
    ✕ Crop Module is added (38ms)
```

HarshKhandeparkar commented:
I think this has got something to do with the `inBrowser` flag not being used in the tests.

HarshKhandeparkar commented:
@jywarren This issue changes 34 files. Isn't it enough for GCI? Fixing the tests etc. wasn't even a part of GCI and I feel it is beyond the scope of GCI. Can my task be approved for now? I can fix the tests separately as a non-GCI issue. Also, this one has taken too much time. It had started about 1.34 weeks ago. I haven't even completed the PR review tasks for the last 2 weeks.

HarshKhandeparkar commented:
@jywarren @SidharthBansal?

keshav234156 commented:
There doesn't seem any issue with UI tests in itself. 
![Screenshot from 2020-01-08 22-29-46](https://user-images.githubusercontent.com/45951376/71998998-9a991c80-3266-11ea-9ce8-e9c3576be13f.png)
Please fix the errors as shown in the console.

HarshKhandeparkar commented:
I don't understand what's broken because the core tests are fine.

On Wed, 8 Jan, 2020, 10:32 PM keshav234156, <notifications@github.com>
wrote:

> There doesn't seem any issue with UI tests in itself.
> [image: Screenshot from 2020-01-08 22-29-46]
> <https://user-images.githubusercontent.com/45951376/71998998-9a991c80-3266-11ea-9ce8-e9c3576be13f.png>
> Please fix the errors as shown in the console.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/image-sequencer/pull/1404?email_source=notifications&email_token=AIJI5H5H3EIM4JCFTSEQ6MTQ4YBITA5CNFSM4KBWFPVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEINIEMQ#issuecomment-572162610>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIJI5H4C7GC7NZFHDCWSEKLQ4YBITANCNFSM4KBWFPVA>
> .
>

jywarren commented:
> I'm not sure what's happening to the tests but I am pretty sure it has got something to do with the new function I added that uses puppeteer to save GIFs in nodejs.

Hi, ok - let's do this: can you turn off your Gif tests for now, but turn them back on in a new branch in a new PR, and describe the issue as well as you can. If we're right, tests here should pass, and we can label GIF support as "experimental" for now until we resolve the issues. Can you update the title to say "experimental GIF support"?

Thank you @HarshKhandeparkar this was a lot of work and I really appreciate it. Let's get pragmatic, close this, and if we want to make a new GCI task for the tests and "graduating" GIF support to non-experimental, we can do that!

Thanks, all!

jywarren commented:
Ah, and we should ensure that the error @keshav234156 pointed [out](https://github.com/publiclab/image-sequencer/pull/1404#issuecomment-572162610) is addressed too before merging!

HarshKhandeparkar commented:
The issue isn't my tests. The UI tests are failing and I have no idea why they are failing. Even if I remove the GIF tests, the UI tests will fail.

HarshKhandeparkar commented:
I don't think I will be able to fix the tests before the deadline. Can the current work be considered for GCI? The fix after that can be non-gci.

I never expected the UI tests to not work. I don't even know what is erroring but it seems like the sequencer in the tests couldn't even start properly.

HarshKhandeparkar commented:
The deadline is tomorrow and I won't be able to work on it today and tomorrow is doubtful(school test and hw). Sorry.

keshav234156 commented on src/modules/_nomodule/PixelManipulation.js:
```diff
@@ -13,9 +20,24 @@
-  const pixelSetter = require('../../util/pixelSetter.js');
-  let wasmSuccess; // Whether wasm succeded or failed
+
+  const isGIF = image.src.includes('gif');
```
this is causing the whole error please change it as it was done previously by you

  const imgType = image.src.split(':')[1].split(';')[0].split('/')[1], // MineType of the image -png, jpg, gif etc.
    isGIF = imgType == 'gif';

HarshKhandeparkar replied:
Why is it causing the error? the image src has to be defined. Why is it not defined?

HarshKhandeparkar commented:
Travis passed
@jywarren @keshav234156

HarshKhandeparkar commented:
I have also added a section about experimental GIF support at the end of the README.

HarshKhandeparkar commented:
I request @jywarren and @keshav234156 to drop a final review. This was an exceptional HoF. Wow! Finally done!

HarshKhandeparkar commented:
Please try to accept my GCI task as soon as possible because I have to complete the last 3 weeks' PR review tasks. Thanks all!

HarshKhandeparkar commented:
@jywarren @keshav234156 I also fixed the module tests which were failing. Not even a single test should be failing now.

keshav234156 commented:
@jywarren i think it is now ready to merge

HarshKhandeparkar commented:
Checklist:
- GIF support for all modules except 6 or 7
- GIFs work in nodejs too
- GIF manipulation is also WASM accelerated both in the browser and node
- A few modules were optimized and work faster now
- Some modules that produced unexpected outputs before are now fixed
- Many module tests that used to fail earlier are all fixed now
- I also fixed the `parseCornerCoordinateInputs` util function which means that inputs like `50%` for width/height or offset are now accepted along with number inputs like `50` for modules like crop etc.
- Added an extra test for GIFs
- The module benchmarks were earlier broken but are now fixed

CC @jywarren @publiclab/is-reviewers

jywarren commented:
This is awesome! An epic work! I'm going through it line by line now. Thanks, Harsh!

HarshKhandeparkar commented:
Thanks a lot!

On Sat, 11 Jan, 2020, 10:13 PM Jeffrey Warren, <notifications@github.com>
wrote:

> This is awesome! An epic work! I'm going through it line by line now.
> Thanks, Harsh!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/image-sequencer/pull/1404?email_source=notifications&email_token=AIJI5H33G25GOMPMEMSTO6LQ5HZMTA5CNFSM4KBWFPVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIWF2EQ#issuecomment-573332754>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIJI5H7TA65B7ZS2FBFXNEDQ5HZMTANCNFSM4KBWFPVA>
> .
>

jywarren commented on examples/demo.js:
```diff
@@ -270,23 +270,26 @@ window.onload = function () {
-
-      // Save the pdf with the filename specified here:
-      pdf.save('index.pdf');
+    sequencer.getImageDimensions(imageDataURL, function(dimensions, isGIF) {
```
Let's mark this for refactoring; we should be able to determine if it's a gif from the imageDataURL directly, with some string parsing. This could be done via a **generalized utility function called `isGIF()` and not required as a parameter**. As long as we clearly mark this it shouldn't be necessary in this PR. Thanks!

jywarren commented on package.json:
```diff
@@ -5,7 +5,7 @@
   "scripts": {
     "debug": "TEST=true node ./index.js -i ./examples/images/monarch.png -s invert",
-    "test": "TEST=true istanbul cover tape test/core/*.js test/core/ui/user-interface.js test/core/modules/*.js | tap-spec; node test/core/sequencer/benchmark.js; grunt tests; cat ./output/core-tests.js | tape-run --render=\"tap-spec\"",
+    "test": "TEST=true istanbul cover tape test/core/*.js test/core/ui/user-interface.js test/core/modules/*.js | tap-spec; node test/core/sequencer/benchmark.js | tap-spec; node test/core/gifs/gif-test.js | tap-spec; grunt tests; cat ./output/core-tests.js | tape-run --render=\"tap-spec\"",
```
Nice!

jywarren commented on src/Modules.js:
```diff
@@ -30,7 +30,8 @@ module.exports = {
   'import-image': require('./modules/ImportImage'),
   'minify-image': require('./modules/MinifyImage'),
-  'invert': require('image-sequencer-invert'),
+  // 'invert': require('image-sequencer-invert'),
```
This had been kept as an example of a module literally living outside the project, which is nice. Maybe we can add a comment back in here to explain this, in follow-up.

jywarren commented on src/modules/Average/Module.js:
```diff
@@ -54,6 +54,7 @@ module.exports = function Average(options, UI) {
     return require('../_nomodule/PixelManipulation.js')(input, {
       output: output,
       ui: options.step.ui,
+      inBrowser: options.inBrowser,
```
Nice!

jywarren commented on src/modules/BlobAnalysis/BlobAnalysis.js:
```diff
@@ -1,17 +1,10 @@
   canvas.height = pixels.shape[1];
   var ctx = canvas.getContext('2d');
-  ctx.drawImage(img[0], 0, 0);
+  ctx.putImageData(new ImageData(new Uint8ClampedArray(pixels.data), pixels.shape[0], pixels.shape[1]), 0, 0);
```
👌

jywarren commented on src/modules/Rotate/Module.js:
```diff
@@ -3,34 +3,57 @@
+      sin = Math.sin(radians);
+      // Final dimensions after rotation
+
+      const finalPixels = require('ndarray')(
```
Thanks for the cleanup!!!!!!

jywarren commented on test/core/modules/add-qr.js:
```diff
@@ -1,5 +1,6 @@
+  image = require('../images/IS-QR'),
   options = {size:200, qrCodeString:'https://github.com/publiclab/image-sequencer'},
-  benchmark = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAAklEQVR4AewaftIAAAApSURBVKXBAQEAAAiAIPP/5+qDMPsIJJJIIokkkkgiiSSSSCKJJJJIogNrygQcXEYsHQAAAABJRU5ErkJggg==';
+  benchmark = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAAEsCAYAAAB5fY51AABlUElEQVR4AezBCXycZ2Hg/9/zvnNqZjQaaaTRfVq2ddiJT8VOHDtOc0FzmIbSpi10IUDaQpajJQnsdmlL6Jb+y/aTljRLSkqBhkINJBA3wYAx2LFjx45v674s6xjNWKPRjDTnO89mwg6I+c9rJ46iSJ+d71fIV5GXl5e3DCjk5eXlLRMKeXl5ecuEQl5eXt4yoZCXl5e3TCjk5eXlLRMKeXl5ecuEQl5eXt4yoZCXl5e3TCjk5eXlLRMKeXl5ecuEQl5eXt4yoZCXl5e3TCjk5eXlLRMKeXl5ecuEQl5eXt4yoZCXl5e3TCjk5e ...
```
Ooh, i dunno -- Let's make a task to turn this into a `require()` so that a) we only have to include it once, and b) it makes these test files more readable. TODO!

jywarren commented:
So the follow-up tasks are:

* [ ] external module definition using `require()` - as an example of a module literally living outside the project (just leave a comment explaining this)
* [ ] refactoring `isGIF()` as a utility function so the parameters don't have to change for the function [here](https://github.com/publiclab/image-sequencer/pull/1404#discussion_r365530741)
* [ ] move the giant dataURLs in the tests into their own file (the same file if they're the same image?) and include them via `require()` to make the actual tests easier to read and reduce file storage (if they are not unique)

Merging the rest, congratulations! 🎉

HarshKhandeparkar commented:
Yeah I'll open the issues. Thanks. Could you please accept my task on the
dashboard too?

On Sat, 11 Jan, 2020, 10:32 PM Jeffrey Warren, <notifications@github.com>
wrote:

> So the follow-up tasks are:
>
>    - external module definition using require() - as an example of a
>    module literally living outside the project (just leave a comment
>    explaining this)
>    - refactoring isGIF() as a utility function so the parameters don't
>    have to change for the function here
>    <https://github.com/publiclab/image-sequencer/pull/1404#discussion_r365530741>
>    - move the giant dataURLs in the tests into their own file (the same
>    file if they're the same image?) and include them via require() to
>    make the actual tests easier to read and reduce file storage (if they are
>    not unique)
>
> Merging the rest, congratulations! 🎉
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/image-sequencer/pull/1404?email_source=notifications&email_token=AIJI5H54COEEY6EVHI7BHKTQ5H3T5A5CNFSM4KBWFPVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIWGG5A#issuecomment-573334388>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIJI5H7NMWENJOGV7DYY7JDQ5H3T5ANCNFSM4KBWFPVA>
> .
>

keshav234156 commented:
@jywarren did we had 0/1 status code enabled for test in the past?

jywarren commented:
I am not sure, but I know that some tests (was it per module tests) were not triggering a Travis failure in the past. Weren't we waiting to resolve the failing tests before re-enabling the status check?

jywarren commented:
Once we refactor the `isGIF()` line, i think we can bump the version number up to the next `v0.x.0` version!

HarshKhandeparkar commented:
I'm not sure I got what you want to say.

On Sat, 11 Jan, 2020, 10:37 PM Jeffrey Warren, <notifications@github.com>
wrote:

> I am not sure, but I know that some tests (was it per module tests) were
> not triggering a Travis failure in the past. Weren't we waiting to resolve
> the failing tests before re-enabling the status check?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/image-sequencer/pull/1404?email_source=notifications&email_token=AIJI5HZ2PX4AFMVJULALABLQ5H4D7A5CNFSM4KBWFPVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIWGJ4A#issuecomment-573334768>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIJI5H6KF4WIRLELNIUO25LQ5H4D7ANCNFSM4KBWFPVA>
> .
>

keshav234156 commented:
@jywarren yes ,can you please have a look at https://github.com/publiclab/image-sequencer/issues/1427#issuecomment-573245999

HarshKhandeparkar commented:
> I am not sure, but I know that some tests (was it per module tests) were not triggering a Travis failure in the past. Weren't we waiting to resolve the failing tests before re-enabling the status check?

Are you talking about travis being broken or the tests? Because I had fixed the tests.

SidharthBansal commented:
This PR is of 1500+ lines. Should we double the points for this task Jeff?

SidharthBansal commented:
I mean 30 points. What do you suggest?

jywarren commented:
:-) sure that sounds great, esp with the tests here!

On Sun, Jan 12, 2020, 9:48 AM Sidharth Bansal <notifications@github.com>
wrote:

> I mean 30 points. What do you suggest?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/image-sequencer/pull/1404?email_source=notifications&email_token=AAAF6JZNNVLOHVFHH3LKX3TQ5MUTFA5CNFSM4KBWFPVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIW3VSQ#issuecomment-573422282>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAF6J7NDPQOKRLPUZRSA43Q5MUTFANCNFSM4KBWFPVA>
> .
>

SidharthBansal commented:
approved 15 points GIF test and 15 points tests.
So total harsh got 30 points.
